typing/ctype.ml

```diff
@@ -5038,69 +5038,103 @@ let rec arity ty =
   | _ -> 0
 
 (* Check for non-generalizable type variables *)
-exception Nongen
-let visited = ref TypeSet.empty
-
-let rec nongen_schema_rec env ty =
-  if TypeSet.mem ty !visited then () else begin
-    visited := TypeSet.add ty !visited;
+let rec nongen_vars_in_schema_fold env ((visited, acc) as acc') ty =
+  if TypeSet.mem ty visited
+  then acc'
+  else begin
+    let visited = TypeSet.add ty visited in
     match get_desc ty with
-      Tvar _ when get_level ty <> generic_level ->
-        raise Nongen
+    | Tvar _ when get_level ty <> generic_level ->
+        visited, TypeSet.add ty acc
     | Tconstr _ ->
-        let old = !visited in
-        begin try iter_type_expr (nongen_schema_rec env) ty
-        with Nongen -> try
-          visited := old;
-          nongen_schema_rec env (try_expand_head try_expand_safe env ty)
-        with Cannot_expand ->
-          raise Nongen
+        let (_, unexpanded_result) as unexpanded_result' =
+          fold_type_expr
+            (nongen_vars_in_schema_fold env)
+            (visited, acc)
+            ty
+        in
+        if unexpanded_result == acc
```

Comment:
A comment that this physical equality is meaningful for sets would be welcome in the future I think.